database/src/migrations/2024062611000000_remove_duplicate_users.ts

```diff
@@ -0,0 +1,249 @@
+    SET system_user_id = dedup.deduplicated_system_user_id
+    FROM DeduplicatedUsers AS dedup
+    WHERE pp.system_user_id IN (
+        SELECT system_user_id
+        FROM system_user
+        WHERE user_identity_source_id = dedup.user_identity_source_id
+        AND LOWER(user_identifier) = LOWER(dedup.user_identifier)
+    )
+    AND pp.system_user_id NOT IN (
+        SELECT deduplicated_system_user_id
+        FROM DeduplicatedUsers
+    );
+
+    -- Deduplicate survey_participation records
+    WITH DeduplicatedUsers AS (
+        SELECT MIN(system_user_id) AS deduplicated_system_user_id,
+        user_identity_source_id,
+        user_identifier
+        FROM system_user
+        GROUP BY user_identity_source_id, user_identifier, record_end_date
+    )
+    UPDATE survey_participation AS pp
+    SET system_user_id = dedup.deduplicated_system_user_id
+    FROM DeduplicatedUsers AS dedup
+    WHERE pp.system_user_id IN (
+        SELECT system_user_id
+        FROM system_user
+        WHERE user_identity_source_id = dedup.user_identity_source_id
+        AND LOWER(user_identifier) = LOWER(dedup.user_identifier)
+    )
+    AND pp.system_user_id NOT IN (
+        SELECT deduplicated_system_user_id
+        FROM DeduplicatedUsers
+    );
+
+    ----------------------------------------------------------------------------------------
+    -- Delete duplicate records
+    ----------------------------------------------------------------------------------------
+
+    -- Remove any duplicate system user role records
+    DELETE FROM system_user_role
+    WHERE system_role_id NOT IN (
+        SELECT MIN(system_user_role_id)
+        FROM system_user_role
+        GROUP BY system_user_id, system_role_id
+    );
+
+    -- Remove any duplicate user records
+    DELETE FROM system_user
+    WHERE system_user_id NOT IN (
+        SELECT MIN(system_user_id)
+        FROM system_user
+        GROUP BY user_identity_source_id, user_identifier, record_end_date
+    );
+
+    -- Remove any duplicate access request records
+    DELETE FROM administrative_activity
+    WHERE administrative_activity_id NOT IN (
+        SELECT MIN(administrative_activity_id)
+        FROM administrative_activity
+        GROUP BY administrative_activity_type_id, reported_system_user_id
+    );
+
+    ----------------------------------------------------------------------------------------
+    -- Add constraints and update views
+    ----------------------------------------------------------------------------------------
+
+    -- Add constraints and update views
+    SET SEARCH_PATH='biohub,biohub_dapi_v1';
+
+    -- Drop views
+    DROP VIEW IF EXISTS biohub_dapi_v1.system_user;
+    DROP VIEW IF EXISTS biohub_dapi_v1.administrative_activity;
+
+    -- Add unique constraint to prevent duplicate users and access requests
+    ALTER TABLE biohub.system_user ADD CONSTRAINT system_user_uk2 UNIQUE (LOWER(user_identifier), user_identity_source_id, record_end_date);
```

Comment:
I'm not sure if LOWER() should be directly in the constraint?

Reply:
Assuming we want it to be case-insensitive, then using LOWER() in the constraint is the best option (for any case-insensitive SQL really).